### DISCUSSION TITLE
Type narrowing after sentinel checks

### DISCUSSION THREAD

**Discussion by aschmolck** [microsoft/pyright#5395] [Q&A] [answered]
One idiom that's sometimes useful in python is to have a sentinel object in places like a function argument default value or as the second argument to `next`. Further it's often useful for this sentinel to be "out-of-type", since the type may be generic or we need to distinguish all members of the type from the sentinel value (e.g. EOF when reading individual bytes).

Here's a concrete example, decorating a stream with a boolean that indicates if the value read is the final value in the stream:

```python
#!/usr/bin/env python3

from typing import Iterable, Iterator, TypeVar

T = TypeVar("T")


def mark_last(it: Iterable[T]) -> Iterator[tuple[bool, T]]:
    it = iter(it)
    sentinel = object()
    x = next(it, sentinel)
    if x is sentinel:
        return
    for subsequent in it:
        yield False, x
        x = subsequent
    yield True, x
```

Unfortuantely that does not typecheck in pyright, because pyright does not narrow the type of x to `T` after the `if x is sentinel: return`. Is there some way to make this sentinel check narrowing work without `cast` or similar?

Note that in this particular case, we can easily enough rewrite the function to not use a sentinel value, and that then type checks (see below), but I'm wondering what to do in other cases where that's not so easily possible.

```python
def mark_last(it: Iterable[T]) -> Iterator[tuple[bool, T]]:
    it = iter(it)
    for last in it:
        break
    else:
        return
    for subsequent in it:
        yield False, last
        last = subsequent
    yield True, last
```

**Comment by erictraut** [comment, marked as answer]
The Python type system is designed to evaluate types, not values. Sentinels are special in that they are singleton objects, but type checkers don't know about this property, and there's no way in the current type system to designate a type as a sentinel. There have been some discussions on the python typing forums to add such a capability to the type system, but these discussions haven't led to any concrete proposals.

There are a few sentinel-like types built in to the stdlib that type checkers know about. This includes `None`, `False`, `True`, and enum values. Leveraging this, you could change your code as follows:

```python
class Sentinel(Enum):
    S = 0


def mark_last(it: Iterable[T]) -> Iterator[tuple[bool, T]]:
    it = iter(it)
    x = next(it, Sentinel.S)
    if x is Sentinel.S:
        return
    for subsequent in it:
        yield False, x
        x = subsequent
    yield True, x
```

Alternatively, you could switch from an `is` type guard to an `isinstance` type guard.

```python
class Sentinel:
    ...

def mark_last(it: Iterable[T]) -> Iterator[tuple[bool, T]]:
    it = iter(it)
    sentinel = Sentinel()
    x = next(it, sentinel)
    if isinstance(x, Sentinel):
        return
    for subsequent in it:
        yield False, x
        x = subsequent
    yield True, x
```

**Reply by aschmolck** [reply]
Thanks Eric, very helpful -- enums as identity-comparable sentinels the typechecker understands was the trick I was missing!
